Issue #13349 - only hard close the websocket connection if abnormal close code

lachlan-roberts opened the pull request:
## Closes #12235

This change avoids hard closing the websocket connection in cases with a normal close code.

This will avoid a race in HTTP/2 where the RST_STREAM frame is racing against the DATA frame holding the WebSocket close.

See comment [#13349](https://github.com/jetty/jetty.project/issues/13349#issuecomment-3252137932) for more details.

lachlan-roberts committed:
Issue #13349 - only hard close the websocket connection if abnormal close code

Signed-off-by: Lachlan Roberts <lachlan.p.roberts@gmail.com>

sbordet commented on jetty-core/jetty-websocket/jetty-websocket-core-common/src/main/java/org/eclipse/jetty/websocket/core/WebSocketCoreSession.java:
```diff
@@ -228,7 +228,16 @@ private void closeConnection(CloseStatus closeStatus, Callback callback)
+        else
+        {
+            connection.cancelDemand();
+            connection.getEndPoint().shutdownOutput();
```
Now the EndPoint is not closed here.
We cannot rely on the other peer to behave properly.
How do we close the EndPoint now?

lachlan-roberts committed:
PR #13557 - ensure EndPoint.close() is always called in normal close case

Signed-off-by: Lachlan Roberts <lachlan.p.roberts@gmail.com>

lachlan-roberts committed:
PR #13557 - integrate close endpoint logic into WebSocketSessionState

Signed-off-by: Lachlan Roberts <lachlan.p.roberts@gmail.com>

lachlan-roberts committed:
PR #13557 - initiate tcp close of the websocket connection from Server

Signed-off-by: Lachlan Roberts <lachlan.p.roberts@gmail.com>

lachlan-roberts committed:
PR #13557 - fix bug in JettyWebSocketFrameHandler

The OnWebSocketClose notification can be called in two cases, if a close frame is received, or if the connection is notified as closed through FrameHandler.onClosed.

The callback should only be succeeded not failed if this is called twice, because this is a normal case.
Received a response close frame and OnWebSocketClose is called through JettyWebSocketFrameHandler.onFrame.
Then the connection is closed so JettyWebSocketFrameHandler.onClosed is called, but we cannot call OnWebSocketClose again.

Signed-off-by: Lachlan Roberts <lachlan.p.roberts@gmail.com>

lachlan-roberts committed:
PR #13557 - ensure endpoint is closed after sessionState.onError(t) is called

Signed-off-by: Lachlan Roberts <lachlan.p.roberts@gmail.com>

lachlan-roberts committed:
PR #13557 - cleanup in WebSocketSessionState

Signed-off-by: Lachlan Roberts <lachlan.p.roberts@gmail.com>

lachlan-roberts committed:
PR #13557 - handle case when close frame fails to be sent

Signed-off-by: Lachlan Roberts <lachlan.p.roberts@gmail.com>

lachlan-roberts commented:
@sbordet please review this PR again.

gregw commented on jetty-core/jetty-websocket/jetty-websocket-core-common/src/main/java/org/eclipse/jetty/websocket/core/internal/WebSocketSessionState.java:
```diff
@@ -37,31 +38,50 @@ enum State
 
-    private final AutoLock lock = new AutoLock();
-    private State _sessionState = State.CONNECTING;
+    enum EndPointState
+    {
+        OPEN,
+        ISHUT,
+        OSHUT,
+        CLOSED
+    }
```
I'm very VERY cautious about this!

Firstly it is duplicating state from inside the endpoint, which itself knows if it is ISHUT or OSHUT, and you can query that state with `endPoint.isInputShutdown()` and `endpoint.isOutputShutdown()`.     Duplicating state is never a good idea and always results in at least small period of incoherent state.

But then I do not think that `EndPointState` is truly an orthogonal state to `WebSocketState`.   Can the endpoint really be in any of those 4 states whilst the session is in any of its states?

gregw reviewed (changes requested):
I'd very much prefer  a single state machine.  Is there a reason that can't be done?
